### PR TITLE
[camera_android] Guard capture-session operations against teardown races

### LINKED ISSUE #82031: [camera] In Android take foto `unlockAutoFocus` error Fatal Exception: java.lang.NullPointerException
## Steps to Reproduce

This issue was experienced by some / many of my users when taking photos of the force close application. And to get this issue just take normal photos on one of the following devices:

1. vivo 1915
2. Oppo A1k
3. vivo 1724
4. realme C15
5. Oppo F7
6. vivo 1915
7. Galaxy A50s

**Expected results:** 

**Actual results:** 

image

<details>
  <summary>Logs</summary>

```
Fatal Exception: java.lang.NullPointerException: Attempt to invoke virtual method 'int android.hardware.camera2.CameraCaptureSession.capture(android.hardware.camera2.CaptureRequest, android.hardware.camera2.CameraCaptureSession$CaptureCallback, android.os.Handler)' on a null object reference
       at io.flutter.plugins.camera.Camera.unlockAutoFocus(Camera.java:26)
       at io.flutter.plugins.camera.Camera.access$1900(Camera.java)
       at io.flutter.plugins.camera.Camera$4.onCaptureCompleted(Camera.java:2)
       at java.lang.reflect.Method.invoke(Method.java)
       at android.hardware.camera2.dispatch.InvokeDispatcher.dispatch(InvokeDispatcher.java:39)
       at android.hardware.camera2.dispatch.HandlerDispatcher$1.run(HandlerDispatcher.java:65)
       at android.os.Handler.handleCallback(Handler.java:790)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:192)
       at android.app.ActivityThread.main(ActivityThread.java:6702)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:549)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:826)
```

```
```

```
```

</details>

### LINKED ISSUE #158878: [camera_android] Crash CameraDevice was already closed `unlockAutoFocus` > `checkIfCameraClosedOrInError`
Seeing this in packages CI as Firebase flakes:

```
java.lang.IllegalStateException: CameraDevice was already closed
     FATAL EXCEPTION: CameraBackground
Process: io.flutter.plugins.cameraexample, PID: 23790
java.lang.IllegalStateException: CameraDevice was already closed
	at android.hardware.camera2.impl.CameraDeviceImpl.checkIfCameraClosedOrInError(CameraDeviceImpl.java:2445)
	at android.hardware.camera2.impl.CameraDeviceImpl.submitCaptureRequest(CameraDeviceImpl.java:1237)
	at android.hardware.camera2.impl.CameraDeviceImpl.capture(CameraDeviceImpl.java:1099)
	at android.hardware.camera2.impl.CameraCaptureSessionImpl.capture(CameraCaptureSessionImpl.java:189)
	at io.flutter.plugins.camera.Camera.unlockAutoFocus(Camera.java:807)
	at io.flutter.plugins.camera.Camera$3.onCaptureCompleted(Camera.java:719)
	at android.hardware.camera2.impl.CameraCaptureSessionImpl$1.lambda$onCaptureCompleted$4$android-hardware-camera2-impl-CameraCaptureSessionImpl$1(CameraCaptureSessionImpl.java:708)
	at android.hardware.camera2.impl.CameraCaptureSessionImpl$1$$ExternalSyntheticLambda1.run(Unknown Source:8)
	at android.os.Handler.handleCallback(Handler.java:942)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.os.HandlerThread.run(HandlerThread.java:67)
```
See link from https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8731360356815402769/+/u/Run_package_tests/Firebase_Test_Lab/stdout

I see this mentioned in comment  https://github.com/flutter/flutter/issues/92762#issuecomment-1408013677 but I _think_ that issue is tracking something different.

### PR DESCRIPTION
**Pull request by kwikwag**
## Summary

Prevents crashes caused by `CameraCaptureSession` being cleared, replaced, or closed while an asynchronous camera operation is still in progress.

This addresses:

* `NullPointerException` in `unlockAutoFocus` when the session is cleared between being checked and used.
* `IllegalStateException` when an operation is submitted to a session whose underlying `CameraDevice` has already closed.
* Stale capture callbacks attempting to operate on a newer capture session.

Fixes flutter/flutter#82031
Fixes flutter/flutter#158878

## Changes

* Snapshot the active capture session and consistently use that snapshot throughout the camera operation, including passing it into helper methods.
* Mark `captureSession` as `volatile` to ensure that updates are visible across the background and platform threads.
* Ignore still-capture completion callbacks from sessions that are no longer current.
* Handle `IllegalStateException` from stale or closed sessions.
* Report an appropriate camera error when a still-capture request cannot continue
* Add regression tests covering null, closed, replaced, and mid-operation session changes.
* Mark `captureFile` as `@VisibleForTesting` for the new video-recording teardown tests.
* Remove `throws` from `Camera.pausePreview` and handle exceptions internally.

Some things I didn't change which might be wanted:

* The AI assistant really wanted to invoke `cameraCaptureCallback.setCameraState(CameraState.STATE_PREVIEW)` on failure conditions so as not to reach bad state. However, existing exception handling didn't reset state, so this is out of scope for what this pull request is trying to fix.
* The responsibility for exception handling is divided inconsistently between `Camera` and `CameraApiImpl`. Specifically this was apparent when applying the fix to `pausePreview()`, as it would be the only function where I wouldn't deal with the exception internally in `Camera`. I decided to match `pausePreview()` with `resumePreview()` in the sense that it handles exceptions internally. Generally, I think exception handling should be made consistent, either in `Camera` or in `CameraApiImpl`. In addition, FlutterError error names should be made more consistent - e.g. `CameraAccessException` sometimes bubbles up under the same name and sometime as `CameraAccess`.
* Some places such as `runPrecaptureSequence` silently swallow exceptions rather than raising it to Flutter. I followed their lead, but it might be wrong.

## Testing

Added Android unit tests covering:

* Session removal between related capture calls.
* Session replacement while an operation is in progress.
* Completion callbacks arriving from stale sessions.
* Closed sessions throwing `IllegalStateException`.
* Null-session handling during still capture and recording.

## Pre-Review Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [AI contribution guidelines] and understand my responsibilities, or I am not using AI tools.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter].
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I followed [the version and CHANGELOG instructions], using [semantic versioning] and the [repository CHANGELOG style], or I have commented below to indicate which documented exception this PR falls under[^1].
- [x] I updated/added any relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or I have commented below to indicate which [test exemption] this PR falls under[^1].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[AI contribution guidelines]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#ai-contribution-guidelines
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[the auto-formatter]: https://github.com/flutter/packages/blob/main/script/tool/README.md#format-code
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[the version and CHANGELOG instructions]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version-and-changelog-updates
[semantic versioning]: https://dart.dev/tools/pub/versioning#semantic-versions
[repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[test exemption]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

### CODE REVIEW

File: packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 1]
```diff
@@ -834,37 +889,68 @@ private void lockAutoFocus() {
...
+              : e.getMessage();
+      dartMessenger.sendCameraErrorEvent(message);
+      return;
     }
 
     refreshPreviewCaptureSession(
+        session,
         null,
         (errorCode, errorMessage) ->
             dartMessenger.error(flutterResult, errorCode, errorMessage, null));
```
[the remark this answers is not in the document]
Comment (kwikwag) [reply packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java:956]: `refreshPreviewCaptureSession` has it's own `try..catch` clause around the `session.setRepeatingRequest()` call, which handles `IllegalStateException` and `CameraAccessException`. It does not handle [`IllegalArgumentException`](https://developer.android.com/reference/java/lang/IllegalArgumentException) but I feel that this could be the subject of some subsequent pull request that deals with exhaustively going over all Camera2 API calls and making sure all exceptions are properly handled and in a consistent manner - in the sense of naming, logging, bubbling up to Flutter, handling internal state, etc.

**Comment by kwikwag** [comment]
Hey @stuartmorgan-g @camsim99 any updates?

**Review by camsim99 — APPROVED** [review]
LGTM! Thank you for the thorough description!

File: packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 2]
```diff
@@ -120,7 +120,13 @@ class Camera
   private HandlerThread backgroundHandlerThread;
 
   CameraDeviceWrapper cameraDevice;
-  CameraCaptureSession captureSession;
+
+  /**
+   * Written from the background handler thread (via {@link CameraCaptureSession.StateCallback}) and
```
Comment (mboetger) [inline packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java:125]: Then this needs to be wrapped in a a synchronized block with a lock variable to make it thread safe.

File: packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java [hunk 3]
```diff
@@ -574,22 +582,33 @@ private void createCaptureSession(
     cameraDevice.createCaptureSession(surfaces, callback, backgroundHandler);
   }
 
-  // Send a repeating request to refresh  capture session.
+  // Send a repeating request to refresh capture session.
   void refreshPreviewCaptureSession(
       @Nullable Runnable onSuccessCallback, @NonNull ErrorCallback onErrorCallback) {
-    Log.i(TAG, "refreshPreviewCaptureSession");
+    // captureSession might be modified in another thread during this method's run
```
Comment (mboetger) [inline packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java:588]: This is not enough to prevent all the ways the concurrency issues could arise - I'd rather wrap things with the synchronized block

**Review by mboetger — CHANGES_REQUESTED** [review]

File: packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java (hunk 3 above)
Comment (kwikwag) [reply packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java:588]: Hey, thanks for your comments. I considered using `synchronized` quite heavily, but opted against it for this pull request. The main reason was that I could not anticipate with the current state of code, that this would not lead to a deadlock. I thought of this pull request as keeping things pretty much the way they are without affecting behavior, except to prevent otherwise unhandled exceptions. It does so by doing everything I could think of *without* resorting to `synchronized` that would mitigate that problem. I think to use `synchronized` would mean also making a slew of tests to see that every place that touches the session and *any other thread-shared member* to synchronized, and thus the scope quickly escalates. I would also mention, that for the most part this plug-in doesn't present that many race conditions in the wild, and I think most instances will be taken care of with these small fixes. Given that, I opted for this solution, and hope we can approve it before making another pull-request to make the plug-in truly thread-safe.

Comment (kwikwag) [reply packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java:588]: @mboetger please let me know if this is satisfactory treatment of your concern.

Comment (kwikwag) [reply packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java:588]: And actually, more concretely - besides the understandable preference towards `synchronized`, is there a specific race condition I have overlooked? I thought the 'store snapshot to reference in a local var' and catch the IllegalStateException pattern is pretty solid.

Comment (mboetger) [reply packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java:588]: I don't think a deadlock is a concern if there is only a single lock object guarding the captureSession.  And yes, without a synchronized block there is still any issue of timing with tearing down the capture session so this does not necessarily fix the problem.  I will take a look at the scope creep and get back to you but I still prefer the synchronized block solution.

Comment (kwikwag) [reply packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java:588]: @mboetger Thank you. I do think this pull request does factually solve *most* if not all of the race cases, and while a `synchronized` *might* be better (I'm not yet convinced it is) I would like to see it merged as I think it brings a much-needed solution for apps that use `startImageStream` and suffer from actual in-the-wild crashes. I was seeing these crashes in my app and they don't happen anymore with this fix. Having said that, I would like to understand better what you mean by "issue of timing with tearing down the capture session". *If* `captureSession` is thread-safe (and AFAIK it is), then the *worst* that can happen is that we get an exception when invoking methods on it. And we handle those exceptions. If I'm missing something, tell me.

Comment (kwikwag) [reply packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java:588]: Hey @mboetger any comments? I honestly don't know if I'll have the time to land another PR for `synchronized` and I feel like it's a pity that this PR, which at the very least improves the code quality as well as fixes some in-the-wild errors, won't get merged just because there might be an "even better" solution.

Comment (mboetger) [reply packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java:588]: Adding the synchronized block does not dramatically increase the size of this PR and is the correct fix. The concerns of deadlocks and scope creep are unfounded.

Comment (kwikwag) [reply packages/camera/camera_android/android/src/main/java/io/flutter/plugins/camera/Camera.java:588]: Ok. Thanks @mboetger . While I understand your wish for proper code, I don't see why not to accept this PR. Having said that, I unfortunately do not have time to further invest in this issue in the foreseeable future. If that changes I'll post another PR, but for now this is the best I can do.

**Comment by kwikwag** [comment]
Also FYI @camsim99 I'm giving up on this PR for the time being. It'd be great if someone else could either complete the work for implementing `syncrhonized` as per @mboetger's suggestion or something else to solve this. Keep me posted.

**Comment by camsim99** [comment]
@kwikwag Thanks for the contribution! Going to close this for now but feel free to send us another PR if you decide to pick this back up.
